### PR TITLE
demo/header-footer.html

### LINKED ISSUE #45: demo/header-footer.html
Demo showcasing a virtual-list with headers/footers

### PR DESCRIPTION
**Pull request by valdrinkoshi**
Fixes #45

### CODE REVIEW

File: demo/header-footer.html [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+  <virtual-list></virtual-list>
+
+  <script type="module">
+    import '../virtual-list-element.js';
+    import './contacts/contact-element.js';
+
+    const virtualList = document.querySelector('virtual-list');
+
+    const nodePool = { group: [], contact: [] };
```
Comment (domenic) [inline demo/header-footer.html]: Add a comment here to explain that our items come in two shapes: contacts, or groups = `{ group: true, title, count }`. So, each of the following functions treats these cases different, and we have two different node pools for recycling among them.

File: demo/header-footer.html [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    virtualList.recycleChild = (element, item) => {
+      if (item.group) {
+        nodePool.group.push(element);
+      } else {
+        nodePool.contact.push(element);
+      }
+    };
+
+    // Get contacts, sort and group by name.
+    fetch('contacts/contacts.json')
```
Comment (domenic) [inline demo/header-footer.html]: Expand the comment here to explain what the input looks like and what the output will look like.

File: demo/header-footer.html [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      .then((resp) => resp.json())
+      .then((contacts) => {
+        let group = null;
+        return contacts.sort((a, b) => {
+          const nameA = a.name.toUpperCase(); // ignore upper and lowercase
```
Comment (domenic) [inline demo/header-footer.html]: You can just do `(a, b) => a.name.localCompare(b.name)`.

File: demo/header-footer.html [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+          if (nameA < nameB) {
+            return -1;
+          }
+          if (nameA > nameB) {
+            return 1;
+          }
+          // names must be equal
+          return 0;
+        }).reduce((items, item) => {
+          let title = item.name.substr(0, 1);
```
Comment (domenic) [inline demo/header-footer.html]: Better is `item.name[0]`.

File: demo/header-footer.html [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+          const nameB = b.name.toUpperCase(); // ignore upper and lowercase
+          if (nameA < nameB) {
+            return -1;
+          }
+          if (nameA > nameB) {
+            return 1;
+          }
+          // names must be equal
+          return 0;
+        }).reduce((items, item) => {
```
Comment (domenic) [inline demo/header-footer.html]: Maybe `item` should be named `contact`?

File: demo/header-footer.html (hunk 2 above)
Comment (domenic) [inline demo/header-footer.html]: It might be better to use async-await for this chain

**Comment by valdrinkoshi** [comment]
Nice tips @domenic, demo got even clearer 👌

**Pull request merged**